Dockerfile

```diff
@@ -1,27 +1,23 @@
-# Node LTS
-FROM node:8.11.3-slim
-
-RUN apt-get update && \
-    apt-get install -y --no-install-recommends \
-	    git=1:2.1.4-2.1+deb8u6 \
-	    python=2.7.9-1 \
-	    make=4.0-8.1 \
-	    gcc=4:4.9.2-2 \ 
-	    g++=4:4.9.2-2 \
-	    libc6-dev=2.19-18+deb8u10 && \
-	  apt-get clean && \
-	  rm -rf /var/lib/apt/lists/*
-
-RUN git clone https://github.com/Financial-Times/polyfill-service.git /polyfill
+FROM node:10.15.3-alpine
 
+RUN apk add --no-cache --update bash
+RUN apk add --no-cache --update --virtual build git python make gcc g++
```

Comment:
Why we do not version-lock here anymore?

Reply:
Cause of cached docker image tags and the added noise. We version lock the image tags in our FROM statements, then trust the upstream maintainers of those images to do the right thing with respect to dependencies.

With version-locked packages you get this:
- Person A sees the breakage, cause he hasn't downloaded the image yet or did a docker reset, so gets a new one
- Person A has a broken setup, cause package-0.0.3 is no longer in available
- Person A updates to the new package (there's no other solution here, there's no testing done here, only manual labor)
- Person B's build is now broken, cause he didn't download the new upstream image but uses a cached version
- Person B then asks whoever is responsible or reverts the change or starts a rant about Docker

Now you *may* get different results for different users (so far haven't seen it, even in Dockerfiles much more involved then this). Then build logs can be examined for version diffs (which apk prints to the console) and can be handled from there.

Reply:
So better break something instead of having something stable? Does not sound right to me.

If you want to have upgraded dependencies, then introduce a service which upgrades those dependencies and makes a PR. Within this PR you will have integration/E2E tests validating the behavior and you just need to click `Approve&Merge` to actually get updated dependencies. This is a much more safer way of dealing with this.

Reply:
The operating system maintainers are diligent when it comes to providing libraries. Much more so than the random pip or npm joe developer.

In my experience the maintainers of some pip or npm package are much more likely to introduce unintended breakage, than operating system maintainers.

For most packages in pip and npm, breakage is just a lolcommit to some github away.

Rely on operating system libs as much as you can. Do not build/cater for the edge cases  like leftpad breaking glibc or vice versa. (if such stuff would be common place, we would switch to a better operating system)